packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts

```diff
@@ -0,0 +1,411 @@
+        pathToBlobSha: new Map(),
+        treesPathToTree: new Map(),
+    };
+
+    // This cache is associated with mapping sha to path for last acked summary. We check this by comparing the lastSummaryProposalHandle
+    // which the driver has with the proposedHandle received in summary context in writeSummaryTree call. If they match, it means the last
+    // uploaded summary got acked. However if the summary is not acked, then we overwrite the "blobTreeDedupCaches" with this cache.
+    private previousBlobTreeDedupCaches: IDedupCaches = {
+        blobShaToPath: new Map(),
+        pathToBlobSha: new Map(),
+        treesPathToTree: new Map(),
+    };
+
+    // Last proposed handle of the uploaded app summary.
+    private lastSummaryProposalHandle: string | undefined;
+
+    constructor(
+        private readonly snapshotUrl: string,
+        private readonly getStorageToken: (options: TokenFetchOptions, name?: string) => Promise<string | null>,
+        private readonly logger: ITelemetryLogger,
+        private readonly epochTracker: EpochTracker,
+        private readonly blobCache: Map<string, IBlob | ArrayBuffer> = new Map(),
+    ) {
+    }
+
+    /**
+     * Builts the caches which will be used for blob deduping.
+     * @param snapshotTree - snapshot tree from which the dedup caches are built.
+     */
+    public async buildCachesForDedup(snapshotTree: api.ISnapshotTree) {
+        const prefixedSnapshotTree = this.addAppPrefixToSnapshotTree(snapshotTree);
+        await this.buildCachesForDedupCore(prefixedSnapshotTree);
+        this.previousBlobTreeDedupCaches = { ...this.blobTreeDedupCaches };
+    }
+
+    /**
+     * Builts the caches which will be used for blob deduping.
+     * @param snapshotTree - snapshot tree from which the dedup caches are built.
+     * @param path - path of the current node evaluated.
+     */
+    private async buildCachesForDedupCore(snapshotTree: api.ISnapshotTree, path: string = ""): Promise<api.ISummaryTree> {
+        assert(Object.keys(snapshotTree.commits).length === 0, "There should not be commit tree entries in snapshot");
+
+        const summaryTree: api.ISummaryTree = {
+            type: api.SummaryType.Tree,
+            tree: {},
+        };
+        for (const [key, value] of Object.entries(snapshotTree.blobs)) {
+            const blobValue = this.blobCache.get(value);
+            assert(blobValue !== undefined, "Blob should exists");
+            const hash = await hashFile(
+                blobValue instanceof ArrayBuffer ?
+                    IsoBuffer.from(blobValue) :
+                        IsoBuffer.from(blobValue.content, blobValue.encoding ?? "utf-8"),
+            );
+            // We are setting the content as empty because we won't use it anywhere.
+            // Instead we will use the hash of the blob from pathToBlobSha cache.
+            summaryTree.tree[key] = {
+                type: api.SummaryType.Blob,
+                content: "",
```

Comment:
I'd remove content here completely, to ensure that if we got it wrong and code later on would try to use content, it would crash. Otherwise error in logic in convertSummaryToSnapshotTree() would result that empty thing to go to summary
In other places we put some special signature (like 'BAAD' when filling in memory that should not be used) to help realize that it's garbage, but here I'd rather force crash then right garbage.

Reply:
done, makes sense.